File: scripts/winmd/platform.py
#!/usr/bin/env python3
"""Windows SDK Platform WinMD collection module.

提供从 Windows SDK API 契约中收集平台 WinMD 引用文件，
以及为 MIDL 编译定位 FoundationContract 元数据目录。
"""

from __future__ import annotations

import xml.etree.ElementTree as ET
from pathlib import Path

from plat_info import BuildError, parse_version, require_dir, require_file, platform_xml_path


# ── 内部辅助 ──────────────────────────────────────────────────


def _parse_contracts(platform_xml: Path) -> list[tuple[str, str]]:
    """从 Platform.xml 中解析 API 契约列表。"""
    try:
        root = ET.parse(platform_xml).getroot()
    except ET.ParseError as exc:
        raise BuildError(f"failed to parse {str(platform_xml)}: {exc}") from exc
    except OSError as exc:
        raise BuildError(f"failed to read {str(platform_xml)}: {exc}") from exc

    contracts: list[tuple[str, str]] = []
    for element in root.iter():
        if element.tag.rsplit("}", 1)[-1] != "ApiContract":
            continue

        name = element.attrib.get("name")
        version = element.attrib.get("version")
        if not name or not version:
            raise BuildError(
                f"ApiContract in {str(platform_xml)} is missing name or version"
            )
        contracts.append((name, version))

    if not contracts:
        raise BuildError(f"no ApiContract entries found in {str(platform_xml)}")
    return contracts


# ── 统一收集接口 ──────────────────────────────────────────────


def collect(winsdk_root: Path, winsdk_version: str) -> list[Path]:
    """收集平台 API 契约对应的 WinMD 引用文件。

    Args:
        winsdk_root: Windows SDK 根目录路径。
        winsdk_version: Windows SDK UAP 版本号。

    Returns:
        API 契约 WinMD 文件的 Path 列表。
    """
    _, platform_xml = platform_xml_path(winsdk_root, winsdk_version)
    contracts = _parse_contracts(platform_xml)
    winmds: list[Path] = []

    for contract_name, contract_version in contracts:
        winmd = (
            winsdk_root
            / "References"
            / winsdk_version
            / contract_name
            / contract_version
            / f"{contract_name}.winmd"
        )
        winmds.append(require_file(winmd, f"platform WinMD {contract_name}"))

    return winmds


# ── 元数据目录发现 ────────────────────────────────────────────


def find_metadata_dir(winsdk_root: Path, winsdk_version: str) -> Path:
    """定位 Windows.Foundation.FoundationContract 元数据目录，供 MIDL 使用。

    Args:
        winsdk_root: Windows SDK 根目录路径。
        winsdk_version: Windows SDK UAP 版本号。

    Returns:
        FoundationContract 最高版本元数据目录的 Path 对象。
    """
    foundation_root = require_dir(
        winsdk_root
        / "References"
        / winsdk_version
        / "Windows.Foundation.FoundationContract",
        "Windows.Foundation.FoundationContract metadata directory",
    )
    versions = [path for path in foundation_root.iterdir() if path.is_dir()]
    if not versions:
        raise BuildError(
            f"no FoundationContract versions found under: {str(foundation_root)}"
        )
    return max(versions, key=lambda path: parse_version(path.name))

File: xmake/modules/winui3/winmd/platform.lua
-- 从 Windows SDK 的 Platform.xml 解析 API 契约，并按 References 布局收集对应 WinMD。
-- SDK 路径发现委托给 winui3.sdk。

local platform_winmds = nil
local metadata_dir = nil

local sdk   = import("winui3.sdk")
local utils = import("utils")

-- Platform.xml 中的 ApiContract 条目足够规整，可用 Lua 模式匹配避免引入 XML 解析依赖。
local function _parse_contracts(platform_xml_path)
    local content=io.readfile(platform_xml_path)

    if not content or #content == 0 then
        raise(string.format("Platform.xml 为空：%s", platform_xml_path))
    end

    local contracts = {}
    for name, version in content:gmatch('ApiContract[^>]-name="([^"]+)"[^>]-version="([^"]+)"') do
        table.insert(contracts, {name = name, version = version})
    end

    -- 兼容 name/version 属性顺序相反的 Platform.xml 条目。
    if #contracts == 0 then
        for name, version in content:gmatch('ApiContract[^>]-version="([^"]+)"[^>]-name="([^"]+)"') do
            table.insert(contracts, {name = version, version = name})
        end
    end

    if #contracts == 0 then
        raise(string.format("Platform.xml 中未找到 ApiContract 条目：%s", platform_xml_path))
    end

    return contracts
end

function collect()
    return platform_winmds
end

-- MIDL /metadata_dir 需要 FoundationContract 的具体版本目录。
function find_metadata_dir()
    return metadata_dir
end

-- 发现 Windows SDK、收集平台 WinMD、定位 FoundationContract 元数据目录。
do
    local sdk_root, sdk_version = sdk.get_sdk_info()

    -- 定位 Platform.xml。
    local platform_xml = path.join(sdk_root, "Platforms", "UAP", sdk_version, "Platform.xml")
    if not os.isfile(platform_xml) then
        local uap_ver = sdk.discover_uap_version(sdk_root)
        if uap_ver then
            sdk_version = uap_ver
            platform_xml = path.join(sdk_root, "Platforms", "UAP", sdk_version, "Platform.xml")
        end
    end

    if not os.isfile(platform_xml) then
        raise(string.format("Windows SDK Platform.xml 不存在：%s", platform_xml))
    end

    -- 解析契约并收集 WinMD。
    local contracts = _parse_contracts(platform_xml)
    local winmds = {}
    for _, contract in ipairs(contracts) do
        local winmd = path.join(sdk_root, "References", sdk_version,
                contract.name, contract.version, contract.name .. ".winmd")
        if os.isfile(winmd) then
            table.insert(winmds, winmd)
        else
            raise(string.format(
                "平台 WinMD 文件不存在：%s\n  契约: %s v%s",
                winmd, contract.name, contract.version))
        end
    end
    table.sort(winmds)

    -- 查找 MIDL /metadata_dir 所需的 FoundationContract 版本目录。
    local foundation_root = path.join(sdk_root, "References", sdk_version,
        "Windows.Foundation.FoundationContract")
    if not os.isdir(foundation_root) then
        raise(string.format("FoundationContract 目录不存在：%s", foundation_root))
    end

    local names   = {}
    local entries = {}
    for _, entry in ipairs(os.dirs(path.join(foundation_root, "*"))) do
        local name = path.filename(entry)
        names[#names + 1] = name
        entries[name] = entry
    end
    local best_name = utils.max_version(names)
    if not best_name then
        raise(string.format("FoundationContract 下无版本目录：%s", foundation_root))
    end

    platform_winmds = winmds
    metadata_dir = entries[best_name]
end
